### ISSUE TITLE
AdminCmsContent - cms category bug with filter, pagination (redirect to cms main)

### ISSUE THREAD

**Issue by mantasl** [PrestaShop/PrestaShop#10399]
** Bug description **
1. Open CMS pages (AdminCmsContent controller);
2. Open CMS category;
3. Try to filter, change page, move to other page, change showing product number. Any action redirect you back to 1. CMS page main. If you click again on cms category you will see filtered content (so it works) but why it redirects to main cms page?

All you do - you back to cms main controller.

** Additionnal information **
PrestaShop version: 1.7.4.2 and 1.7.3.3 tested both - same bug;
PHP version: 7.1

---
#### Describe the bug
Any action redirect you back to 1. CMS page main. If you click again on cms category you will see filtered content (so it works) but why it redirects to main cms page?

#### Expected behavior
When we click cancel it should redirect us to Cateory Parent not the Home category

#### Steps to reproduce
1. Create a new page category
2. Click on the view to see pages for this category
3. Create more than two pages
4. Filter pages by Title
5. Edit the page
6. Click on the cancel button, you are redirected to the main page (Design => Pages)
7. the filter is still active

**Screenshots**

https://drive.google.com/file/d/1KXwDh24kdo3sivoKR28_GiUvk0mxLgF7/view

#### Additional information

* PrestaShop version: 1.7.7.1
* PHP version: 7.2

**Comment by khouloudbelguith** [comment]
Hi, @mantasl,

I manage to reproduce the issue with the last version 1.7.4.2.
We will see how to fix it.
Related to #9825, #9853

Best regards, Khouloud

**Comment by MathiasReker** [comment]
I did not manage to solve this, but I think I might have found some useful for a solution.

File: `AdminCmsContentController`

If I change this line 45: `$id_cms_category = (int)Tools::getValue('id_cms_category', Tools::getValue('id_cms_category_parent', 1));`

to

`$id_cms_category = (int)Tools::getValue('id_cms_category', Tools::getValue('id_cms_category_parent', 2));`

it solves the reported issue, but now it redirect from any category to category 2.

Maybe this value should depends on what category you are in when you filter etc.?

**Comment by mantasl** [comment]
Yes, but the problem somewhere else. If you check what you get after u change page / items per page - controller get id cms:1. So after we change something its redirecting to wrong place and after it - get wrong ID and load wrong content.

**Comment by colinegin** [comment]
Hello @khouloudbelguith do you reproduce this bug with the latest version ?

**Comment by khouloudbelguith** [comment]
Hi @colinegin,

I manage to reproduce the issue with PS1.7.6.1.
Steps to reproduce the issue:
1. Create a new page category
2. Click on the view to see pages for this category
3. Create more than two pages
4. Filter pages by Title
5. Edit the page
6. Click on the cancel button, you are redirected to the main page (Design => Pages)
7. the filter is still active

https://drive.google.com/file/d/1KXwDh24kdo3sivoKR28_GiUvk0mxLgF7/view

Thanks!

**Comment by Rubben1985** [comment]
Here is the solution for TB fork (&PS 1.6?)
https://github.com/thirtybees/thirtybees/commit/844f9ec4d028d385c90f192c6b14a020165d54b6

I tried to apply it to PS 1.7 but files are differents.

**Comment by denys202** [comment]
One of the possbile solutions that works is to add the next code, but I am not sure that it a best place to add this code :
```
$this->requestStack->getCurrentRequest()->attributes->set('_route_params', array_merge(
    [
        'id_cms_category' => $cmsCategoryId,
    ],
    $this->requestStack->getCurrentRequest()->attributes->get('_route_params')
));
```
in this file `src/PrestaShopBundle/EventListener/FilterCmsPageCategorySearchCriteriaListener.php`
after the `id_cms_category_parent` assignement: `$filters['id_cms_category_parent'] = $cmsCategoryId;`

**Comment by djbuch** [comment]
@denys202 your solution works for me too.

Thanks !

**Comment by Rubben1985** [comment]
Yes it works! Thanks @denys202

**Comment by boo-code** [comment]
Checked on 9.2. The two halves of this thread have different answers.

**The filter and pagination half no longer reproduces.** The legacy `AdminCmsContentController` that @MathiasReker pointed at is gone, the page is migrated, and the replacement keeps the category across a filter:

```php
return $this->buildSearchResponse(
    $this->container->get($filterId),
    $request,
    $filterId,
    'admin_cms_pages_index',
    [
        'id_cms_category',
    ]
);
```

**The navigation half does reproduce**, matching the expectation stated later in the thread, "it should redirect us to Category Parent not the Home category". Inside a page category, "Back to list" is rendered with no route parameters at all:

```twig
<a href="{{ path('admin_cms_pages_index') }}" class="btn btn-outline-secondary back-to-list-link">
```

so it always lands on the root. The same widget on Sell > Catalog > Categories climbs one level:

```twig
{% if currentCategoryView.id_parent %}
  {% set params = {id_category: currentCategoryView.id_parent} %}
{% endif %}
```

The CMS template had nothing to link to, because `CmsPageViewDataProvider::getView()` returned only `root_category_id` and `breadcrumb_tree`. #42328 adds `parent_category_id`, resolved through the `GetCmsPageParentCategoryIdForRedirection` query the controller already uses for its post-action redirects, and points the link at it.
